sympy/core/sympify.py

```diff
@@ -90,7 +91,7 @@ def _convert_numpy_types(a, **sympify_args):
 
 
 def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
-        evaluate=None):
+        evaluate=None) -> Basic:
```

Comment:
One of the places I got stuck when previously looking at type-hinting is that there is not an accurate way to describe the type that is returned by sympify which is Basic or a list of Basic or a list of list of Basic...
```python
In [1]: sympify([[[1, 2]], [0]])
Out[1]: [[[1, 2]], [0]]
```

Reply:
That is a bit of an issue... the best compromise I can think of is to return something like `Union[Basic, Any]`. I'll investigate this further though... I'm sure there's a solution somewhere.

Reply:
I know very little about Python typing, but I suspect things like this can be solved with parametric types, e.g., `f(x: T) -> T`, where `T = TypeVar("T")` is a function that returns the same type as its input. See https://docs.python.org/3/library/typing.html#typing.TypeVar.